File: examples/python/vamana_index_example.py
"""Vamana index -- DiskANN-style graph. Requires training data."""

import random
import uuid

from membrain import MembrainVamanaIndex

DIMENSION = 128
NUM_TRAINING = 500


def embed(seed: float) -> list[float]:
    random.seed(seed)
    return [random.gauss(0, 1) for _ in range(DIMENSION)]


def main() -> None:
    ids = [str(uuid.uuid4()) for _ in range(NUM_TRAINING)]
    vectors = [embed(seed=float(i)) for i in range(NUM_TRAINING)]

    config = {"max_degree": 64, "alpha": 1.2, "search_list_size": 100, "distance_metric": "Cosine"}

    with MembrainVamanaIndex.build(
        dimension=DIMENSION, ids=ids, vectors=vectors, config=config,
    ) as index:
        print(f"Built Vamana index: {len(index)} vectors")

        results = index.search(embed(seed=0.5), k=5)
        for result in results:
            print(f"  id={result.id[:8]}... score={result.score:.4f}")


if __name__ == "__main__":
    main()

File: examples/javascript/vamana_index_example.mjs
/** Vamana index -- DiskANN-style graph. Requires training data. */

import { randomUUID } from "node:crypto";
import { MembrainVamanaIndex } from "membrain";

const DIMENSION = 128;
const NUM_TRAINING = 500;

function embed(seed) {
  const out = [];
  let state = seed * 1000 + 1;
  for (let i = 0; i < DIMENSION; i++) {
    state = (state * 1103515245 + 12345) & 0x7fffffff;
    out.push((state / 0x7fffffff) * 2 - 1);
  }
  return out;
}

const ids = Array.from({ length: NUM_TRAINING }, () => randomUUID());
const vectors = Array.from({ length: NUM_TRAINING }, (_, i) => embed(i));
const config = { dimension: DIMENSION, max_degree: 64, alpha: 1.2, search_list_size: 100, distance_metric: "Cosine" };

const index = MembrainVamanaIndex.build(ids, vectors, config);

try {
  console.log(`Built Vamana index: ${index.len()} vectors`);

  const results = index.search(embed(0.5), 5);
  for (const r of results) {
    console.log(`  id=${r.id.slice(0, 8)}... score=${r.score.toFixed(4)}`);
  }
} finally {
  index.close();
}
